added permission to get nodes for rbd

nemcikjan opened the pull request:
<!-- Please take a look at our [Contributing](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#Code-contribution-workflow)
documentation before submitting a Pull Request!
Thank you for contributing to ceph-csi! -->

# Describe what this PR does #

Based on the implementation in rbd driver the nodeplugin should have by default permission to get nodes in order to retrieve the node labels

## Is there anything that requires special attention ##

* Related rbd driver implementation can be found [here](https://github.com/ceph/ceph-csi/blob/devel/internal/rbd/driver/driver.go#L128C1-L133C3)

## Future concerns ##

List items that are not part of the PR and do not impact it's
functionality, but are work items that can be taken up subsequently.

**Checklist:**

* [x] **Commit Message Formatting**: Commit titles and messages follow
  guidelines in the [developer
  guide](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#commit-messages).
* [x] Reviewed the developer guide on [Submitting a Pull
  Request](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#development-workflow)

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

* `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)

</details>

Commit:
helm: update node plugin cluster role

added permission to get nodes for rbd and cephfs nodeplugin daemonset

Signed-off-by: Jan Nemcik <jan.nemcik@solargis.com>

Rakshith-R commented:
Thanks for the pr ! 

@iPraveenParihar after the new changes to obtain node labels if the pods are running in k8s cluster, we'll need rbac to get nodes regardless of topology or read affinity right ?

Rakshith-R commented:
@nemcikjan Can you remove the surrounding conditions here too ? 
https://github.com/ceph/ceph-csi/blob/ec05ae6e5f096c5f8be990968f950a6915152b3e/charts/ceph-csi-cephfs/templates/nodeplugin-clusterrole.yaml#L14

nemcikjan commented:
> @nemcikjan Can you remove the surrounding conditions here too ?
> 
> https://github.com/ceph/ceph-csi/blob/ec05ae6e5f096c5f8be990968f950a6915152b3e/charts/ceph-csi-cephfs/templates/nodeplugin-clusterrole.yaml#L14

@Rakshith-R I can but I looked also into the cephfs driver implementation and got the impression that it's not necessary for the cephfs rbac but it's possible that I missed something

Rakshith-R commented:
> > @nemcikjan Can you remove the surrounding conditions here too ?
> > https://github.com/ceph/ceph-csi/blob/ec05ae6e5f096c5f8be990968f950a6915152b3e/charts/ceph-csi-cephfs/templates/nodeplugin-clusterrole.yaml#L14
> 
> @Rakshith-R I can but I looked also into the cephfs driver implementation and got the impression that it's not necessary for the cephfs rbac but it's possible that I missed something

Yes, we're only checking if cephcsi is running inside k8s before trying to fetch the node labels now. 
Therefore, That rbac is needed regardless of the read affinity condition

iPraveenParihar commented:
> Thanks for the pr !
> 
> @iPraveenParihar after the new changes to obtain node labels if the pods are running in k8s cluster, we'll need rbac to get nodes regardless of topology or read affinity right ?

yes, we can remove the read affinity condition.

XtremeOwnageDotCom commented:
Just a note, I also had to update the provisioner role, in my cluster. It apparently also needs the ability to get nodes.

https://github.com/ceph/ceph-csi/blob/devel/charts/ceph-csi-rbd/templates/provisioner-clusterrole.yaml

The exact same resolution, would apply to it.

nemcikjan commented:
@XtremeOwnageDotCom I reviewed the implementation and it doesn't seem that the rbd provisioner needs that permission at all based on the condition [here](https://github.com/ceph/ceph-csi/blob/devel/internal/rbd/driver/driver.go#L142C1-L142C1). Maybe @Rakshith-R or @iPraveenParihar could confirm.

iPraveenParihar commented:
> Just a note, I also had to update the provisioner role, in my cluster. It apparently also needs the ability to get nodes.
> 
> https://github.com/ceph/ceph-csi/blob/devel/charts/ceph-csi-rbd/templates/provisioner-clusterrole.yaml
> 
> The exact same resolution, would apply to it.



In RBD, we need to add the NodeServer condition as done in CephFS
https://github.com/ceph/ceph-csi/blob/23091689434a39317e4790e2312cb8fcf19800dc/internal/rbd/driver/driver.go#L128-L133

CephFS,
https://github.com/ceph/ceph-csi/blob/23091689434a39317e4790e2312cb8fcf19800dc/internal/cephfs/driver.go#L112-L118

Rakshith-R reviewed (commented):
Looks good to me, 
Can you please squash both the commits into one ?

Rakshith-R commented on internal/rbd/driver/driver.go:
```diff
@@ -125,25 +125,21 @@ func (r *Driver) Run(conf *util.Config) {
 	r.ids = NewIdentityServer(r.cd)
 
-	if conf.IsNodeServer {
+	if conf.IsNodeServer && k8s.RunsOnKubernetes() {
 		topology, err = util.GetTopologyFromDomainLabels(conf.DomainLabels, conf.NodeID, conf.DriverName)
 		if err != nil {
 			log.FatalLogMsg(err.Error())
 		}
+		nodeLabels, err = k8s.GetNodeLabels(conf.NodeID)
+		if err != nil {
+			log.FatalLogMsg(err.Error())
+		}
+		if conf.EnableReadAffinity {
+			crushLocationMap = util.GetCrushLocationMap(conf.CrushLocationLabels, nodeLabels)
+		}
```
```suggestion
	if conf.IsNodeServer {
		topology, err = util.GetTopologyFromDomainLabels(conf.DomainLabels, conf.NodeID, conf.DriverName)
		if err != nil {
			log.FatalLogMsg(err.Error())
		}
		if k8s.RunsOnKubernetes() {
		        nodeLabels, err = k8s.GetNodeLabels(conf.NodeID)
		        if err != nil {
			        log.FatalLogMsg(err.Error())
		        }
		}
		if conf.EnableReadAffinity {
			crushLocationMap = util.GetCrushLocationMap(conf.CrushLocationLabels, nodeLabels)
		}
```

The current change will prevent nodeserver from running in non-K8s environment.
Let's check runsOnK8s only while getting node labels.

nemcikjan replied:
@Rakshith-R Actually, this is tricky, because `util.GetTopologyFromDomainLabels` internally also fetches node labels using `k8s.GetNodeLabels` that's why I merged the condition. I suggest to extend `util.GetTopologyFromDomainLabels` by new parameter nodeLabels so we can separately check if the controller is running in k8s and fetch the node labels and then just reuse them

Rakshith-R replied:
> @Rakshith-R Actually, this is tricky, because `util.GetTopologyFromDomainLabels` internally also fetches node labels using `k8s.GetNodeLabels` that's why I merged the condition. I suggest to extend `util.GetTopologyFromDomainLabels` by new parameter nodeLabels so we can separately check if the controller is running in k8s and fetch the node labels and then just reuse them

@nemcikjan 
We'll now be having nodes get rbac by default for nodeserver so let's just make sure GetNodeLabels is covered inside conf.nodeserver struct for backport.

We can make this enhancement in follow-up pr.

nemcikjan replied:
@Rakshith-R OK, I'll make appropriate changes to the scope of this PR

Rakshith-R replied:
> Actually, this is tricky, because `util.GetTopologyFromDomainLabels` internally also fetches node labels using `k8s.GetNodeLabels` that's why I merged the condition. I suggest to extend `util.GetTopologyFromDomainLabels` by new parameter nodeLabels so we can separately check if the controller is running in k8s and fetch the node labels and then just reuse them

That part should be fine since we are now adding rbac always for nodeserver.
let's just make this change to keep GetNodeLabels inside conf.nodeserver  for backports.
We'll make that enhancement in a follow up pr. 

Please check the commitlint failure too 
https://github.com/ceph/ceph-csi/actions/runs/7126377653/job/19405449656?pr=4302

Thanks

Rakshith-R replied:
@nemcikjan ^

Rakshith-R commented:
> Looks good to me, Can you please squash both the commits into one ?

The last commit came after I pressed enter for this review. 
just one more change on the latest commit. 

Thanks

nemcikjan commented:
@Rakshith-R should I then squash all the commits?

Rakshith-R commented:
> @Rakshith-R should I then squash all the commits?

Just the first two since they are so similar
and modify the third commit according to https://github.com/ceph/ceph-csi/pull/4302#discussion_r1418683065

Rakshith-R reviewed (dismissed):
Thanks

Rakshith-R commented on internal/rbd/driver/driver.go:
```diff
@@ -125,17 +125,6 @@ func (r *Driver) Run(conf *util.Config) {
-		crushLocationMap = util.GetCrushLocationMap(conf.CrushLocationLabels, nodeLabels)
-	}
-
 	// Create GRPC servers
```
```suggestion
	if k8s.RunsOnKubernetes() && conf.IsNodeServer {
		nodeLabels, err = k8s.GetNodeLabels(conf.NodeID)
		if err != nil {
			log.FatalLogMsg(err.Error())
		}
	}
	if conf.EnableReadAffinity {
		crushLocationMap = util.GetCrushLocationMap(conf.CrushLocationLabels, nodeLabels)
	}
	
	// Create GRPC servers
```


@nemcikjan  golangci lint is complaining that the below nest if is too complex
https://github.com/ceph/ceph-csi/actions/runs/7139935245/job/19446163983?pr=4302

Let's just add conf.NodeServer check in the same place.

Rakshith-R reviewed (commented):
ci failure https://github.com/ceph/ceph-csi/actions/runs/7139935245/job/19446163983?pr=4302

Commit:
rbd: updated node labels fetching logic

node labels are fetched only if controller is running in k8s and is nodeserver

Signed-off-by: Jan Nemcik <jan.nemcik@solargis.com>

Rakshith-R reviewed (approved):
Thanks !

Rakshith-R commented:
@mergifyio rebase
@mergifyio queue

Rakshith-R commented:
@mergifyio queue
